### PR TITLE
Fix password input

### PR DESCRIPTION
**Pull request by mmso**

### CODE REVIEW

File: components/input/PasswordInput.js [hunk 1]
```diff
@@ -1,24 +1,52 @@
...
+    const { handlers, statusClasses, status } = useInput(props);
+    const [uid] = useState(generateUID('passwordInput'));
 
     return (
         <>
-            <Input disabled={disabled} type={type} {...rest} />
-            <button disabled={disabled} type="button" onClick={toggle}>
-                <Icon name={type === 'password' ? 'read' : 'unread'} />
-            </button>
+            <span className="relative">
```
Comment (EpokK) [inline components/input/PasswordInput.js]: It should be a `div` no?

File: components/input/PasswordInput.js [hunk 2]
```diff
@@ -1,24 +1,52 @@
...
+                <input
+                    className={`pm-field w100 ${className} ${statusClasses}`}
+                    aria-invalid={error && status.isDirty}
+                    aria-describedby={uid}
+                    type={type}
+                    {...rest}
+                    {...handlers}
+                />
+                <button
+                    title={c('Label').t`Reveal password`}
```
Comment (EpokK) [inline components/input/PasswordInput.js]:
```suggestion
                    title={type === 'password' ? c('Label').t`Reveal password` : c('Label').t`Hide password`}
```

File: components/input/PasswordInput.js (hunk 1 above)
Comment (mmso) [reply components/input/PasswordInput.js]: Checked with @nico3333fr and it should be a `span`

File: components/input/useInput.js [hunk 3]
```diff
@@ -1,35 +1,98 @@
...
-const FOCUSSED_CLASS = 'focussed';
-const BLURRED_CLASS = 'blured';
+const FOCUSSED_CLASS = 'focused';
+const BLURRED_CLASS = 'blurred';
 const DIRTY_CLASS = 'dirty';
 const PRISTINE_CLASS = 'pristine';
 
-const DEFAULT_STATE = { focussed: false, blurred: false, dirty: false, pristine: true };
+const DEFAULT_STATE = {
+    isFocused: false,
```
Comment (EpokK) [inline components/input/useInput.js]:
```suggestion
    isFocussed: false,
```

File: components/input/useInput.js [hunk 4]
```diff
@@ -1,35 +1,98 @@
...
+};
 
-const useInput = (initialState = DEFAULT_STATE, prefix = 'field') => {
+const useInput = (props, initialState = DEFAULT_STATE, prefix = 'field') => {
     const [status, changeStatus] = useState(initialState);
-    const change = () => !status.dirty && changeStatus({ ...status, dirty: true, pristine: false });
-    const blur = () => !status.blurred && changeStatus({ ...status, blurred: true, pristine: false });
-    const focus = () => !status.focussed && changeStatus({ ...status, focussed: true, pristine: false });
-    const reset = () => changeStatus({ ...DEFAULT_STATE });
+    const { isFocused, isBlurred, isPristine, isDirty } = status;
```
Comment (EpokK) [inline components/input/useInput.js]:
```suggestion
    const { isFocussed, isBlurred, isPristine, isDirty } = status;
```

File: components/input/useInput.js [hunk 5]
```diff
@@ -1,35 +1,98 @@
...
     const classes = [];
 
-    if (status.pristine) {
+    if (isPristine) {
         classes.push(PRISTINE_CLASS);
     } else {
-        status.focussed && classes.push(`${prefix}-${FOCUSSED_CLASS}`);
-        status.blurred && classes.push(`${prefix}-${BLURRED_CLASS}`);
-        status.dirty && classes.push(`${prefix}-${DIRTY_CLASS}`);
+        isFocused && classes.push(`${prefix}-${FOCUSSED_CLASS}`);
```
Comment (EpokK) [inline components/input/useInput.js]:
```suggestion
        isFocussed && classes.push(`${prefix}-${FOCUSSED_CLASS}`);
```

File: components/input/useInput.js [hunk 6]
```diff
@@ -1,35 +1,98 @@
...
-        focus,
-        reset
+        reset,
+        handlers: {
+            onFocus: (event) => {
+                if (disabled) {
+                    return;
+                }
+
+                if (!status.isFocused) {
```
Comment (EpokK) [inline components/input/useInput.js]:
```suggestion
                if (!status.isFocussed) {
```

File: components/input/useInput.js [hunk 7]
```diff
@@ -1,35 +1,98 @@
...
+                    changeStatus({
+                        ...status,
+                        isFocused: true,
```
Comment (EpokK) [inline components/input/useInput.js]:
```suggestion
                        isFocussed: true,
```

File: components/input/Input.js [hunk 8]
```diff
@@ -1,84 +1,28 @@
...
+                aria-invalid={error && status.isDirty}
                 aria-describedby={uid}
-                onFocus={handleFocus}
-                onBlur={handleBlur}
-                onKeyDown={handleKeyDown}
-                onChange={handleChange}
-                ref={inputRef}
+                ref={ref}
                 {...rest}
+                {...handlers}
```
Comment (EpokK) [inline components/input/Input.js]: This is so much better 👏

File: components/input/PasswordInput.js [hunk 9]
```diff
@@ -1,24 +1,52 @@
...
-            <button disabled={disabled} type="button" onClick={toggle}>
-                <Icon name={type === 'password' ? 'read' : 'unread'} />
-            </button>
+            <span className="relative">
+                <input
+                    className={`pm-field w100 ${className} ${statusClasses}`}
+                    aria-invalid={error && status.isDirty}
+                    aria-describedby={uid}
+                    type={type}
+                    {...rest}
```
Comment (EpokK) [inline components/input/PasswordInput.js]: Maybe `{...rest}` should be defined after `{...handlers}` to let the dev override it?

Comment (mmso) [reply components/input/PasswordInput.js]: No because useInput needs to take control of it, it'll be called if it's defined inside there

**Pull request merged**